docs/proposals/tenant_retention.md

```diff
@@ -0,0 +1,64 @@
+---
+title: "Retention of Tenant Data from Blocks Storage"
+linkTitle: "Retention of Tenant Data from Blocks Storage"
+weight: 1
+slug: tenant-retention
+---
+
+- Author: [Allenzhli](https://github.com/Allenzhli)
+- Date: December 2020
+- Status: Proposed
+
+## Retention of tenant data
+
+## Problem
+
+In Cortex, we are missing retention of blocks storage. The metric data is growing over time per-tenant, at the same time, the value of data decreases. We want to support retention time of metric data like prometheus does.
+
+## Background
+
+When using blocks storage, Cortex stores tenant’s data in object store for long-term storage of blocks, tenant id as part of the object store path. We discover all tenants via scan the root dir of bucket.
+
+## Proposal
+
+### API Endpoints
+
+#### POST /compactor/retention_tenant
+
+We propose to introduce an `/compactor/retention_tenant` API endpoint to set data retention for the tenant.
+
+While this endpoint works as an “admin” endpoint and should not be exposed directly to tenants, it still needs to know under which tenant it should operate.
+For consistency with other endpoints this endpoint would therefore require an `X-Scope-OrgID` header and use the tenant from it.
+
+It is safe to call “retention_tenant” multiple times.
+
+#### GET /compactor/retention_tenant
+
+To query the state of data retention, an endpoint will be available: `/compactor/retention_tenant`.
+This will return the retention of current tenant.
+For consistency with other endpoints this endpoint would therefore require an `X-Scope-OrgID` header and use the tenant from it.
+
+### Implementation (asynchronous)
+
+Compactor will implement both API endpoints.
+Upon receiving the call to `/compactor/retention_tenant` endpoint, Compactor will initiate the retention by writing “retetion marker” objects for specific tenant to blocks bucket.
+
+Retention marker for the tenant will be an object stored under tenant prefix, eg. “/<tenant>/retention”.
```

Comment:
I wouldn't use any retention marker. The per-tenant retention should be part of the config.

Reply:
I agree with you, this will be delete.